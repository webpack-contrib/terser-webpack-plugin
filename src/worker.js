import minify from './minify';

module.exports = (options, callback) => {
  try {
    // 'use strict' => this === undefined (Clean Scope)
    // Safer for possible security issues, albeit not critical at all here
    // eslint-disable-next-line no-new-func, no-param-reassign
    options = new Function(
      'exports',
      'require',
      'module',
      '__filename',
      '__dirname',
      `'use strict'\nreturn ${options}`
    )(exports, require, module, __filename, __dirname);

    callback(null, minify(options));
  } catch (errors) {
    callback(errors);
  }
};
