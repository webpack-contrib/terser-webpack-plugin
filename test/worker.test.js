import serialize from 'serialize-javascript';

import worker from '../src/worker';

describe('matches snapshot', () => {
  it('normalizes when options.extractComments is regex', () => {
    const options = {
      file: 'test1.js',
      input: 'var foo = 1;/* hello */',
      extractComments: /foo/,
    };
    worker(serialize(options), (error, data) => {
      if (error) {
        throw error;
      }
      expect(data).toMatchSnapshot(options.file);
    });
  });

  it('normalizes when terserOptions.output.comments is string: all', () => {
    const options = {
      file: 'test2.js',
      input: 'var foo = 1;/* hello */',
      terserOptions: {
        output: {
          comments: 'all',
        },
      },
    };
    worker(serialize(options), (error, data) => {
      if (error) {
        throw error;
      }
      expect(data).toMatchSnapshot(options.file);
    });
  });

  it('normalizes when terserOptions.output.comments is string: some', () => {
    const options = {
      file: 'test3.js',
      input: 'var foo = 1;/* hello */',
      terserOptions: {
        output: {
          comments: 'some',
        },
      },
    };
    worker(serialize(options), (error, data) => {
      if (error) {
        throw error;
      }
      expect(data).toMatchSnapshot(options.file);
    });
  });

  it('normalizes when terserOptions.extractComments is number', () => {
    const options = {
      file: 'test4.js',
      input: 'var foo = 1;/* hello */',
      terserOptions: {
        output: {
          comments: 'some',
        },
      },
      extractComments: 1,
    };
    worker(serialize(options), (error, data) => {
      if (error) {
        throw error;
      }
      expect(data).toMatchSnapshot(options.file);
    });
  });

  it('when applied with extract option set to a single file', () => {
    const options = {
      file: 'test5.js',
      input: '/******/ function hello(a) {console.log(a)}',
      terserOptions: {
        output: {
          comments: 'all',
        },
      },
      extractComments: {
        condition: 'should be extracted',
        filename: (file) => file.replace(/(\.\w+)$/, '.license$1'),
        banner: (licenseFile) =>
          `License information can be found in ${licenseFile}`,
      },
    };
    worker(serialize(options), (error, data) => {
      if (error) {
        throw error;
      }
      expect(data).toMatchSnapshot(options.file);
    });
  });

  it('when options.inputSourceMap', () => {
    const options = {
      file: 'test6.js',
      input: 'function foo(x) { if (x) { return bar(); not_called1(); } }',
      inputSourceMap: {
        version: 3,
        sources: ['test1.js'],
        names: ['foo', 'x', 'bar', 'not_called1'],
        mappings: 'AAAA,QAASA,KAAIC,GACT,GAAIA,EAAG,CACH,MAAOC,MACPC',
      },
    };
    worker(serialize(options), (error, data) => {
      if (error) {
        throw error;
      }
      expect(data).toMatchSnapshot(options.file);
    });
  });
});
