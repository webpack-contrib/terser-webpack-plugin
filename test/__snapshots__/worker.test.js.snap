// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot normalizes when options.extractComments is regex: test1.js 1`] = `
Object {
  "code": "var foo=1;",
  "error": undefined,
  "extractedComments": Array [],
  "map": undefined,
  "warnings": undefined,
}
`;

exports[`matches snapshot normalizes when terserOptions.extractComments is number: test4.js 1`] = `
Object {
  "code": "var foo=1;",
  "error": undefined,
  "extractedComments": Array [],
  "map": undefined,
  "warnings": undefined,
}
`;

exports[`matches snapshot normalizes when terserOptions.output.comments is string: all: test2.js 1`] = `
Object {
  "code": "var foo=1;/* hello */",
  "error": undefined,
  "extractedComments": Array [],
  "map": undefined,
  "warnings": undefined,
}
`;

exports[`matches snapshot normalizes when terserOptions.output.comments is string: some: test3.js 1`] = `
Object {
  "code": "var foo=1;",
  "error": undefined,
  "extractedComments": Array [],
  "map": undefined,
  "warnings": undefined,
}
`;

exports[`matches snapshot when applied with extract option set to a single file: test5.js 1`] = `
Object {
  "code": "/******/function hello(o){console.log(o)}",
  "error": undefined,
  "extractedComments": Array [],
  "map": undefined,
  "warnings": undefined,
}
`;

exports[`matches snapshot when options.inputSourceMap: test6.js 1`] = `
Object {
  "code": "function foo(f){if(f)return bar()}",
  "error": undefined,
  "extractedComments": Array [],
  "map": "{\\"version\\":3,\\"sources\\":[\\"test1.js\\"],\\"names\\":[\\"foo\\",\\"x\\",\\"not_called1\\"],\\"mappings\\":\\"AAAA,SAASA,IAAIC,GACT,GACIA,EAAA,OACAC\\"}",
  "warnings": undefined,
}
`;
